### PR TITLE
FEATURE: Ability to dismiss all new topics

### PR DESCRIPTION
**Pull request by lis2**
Follow up https://github.com/discourse/discourse/pull/11968

Dismiss all new topics using the same DismissTopicService. In addition, MessageBus receives exact topic ids which should be marked as `seen`.

I added a few comments to make it easier to review

### CODE REVIEW

File: app/assets/javascripts/discourse/app/models/topic-tracking-state.js [hunk 1]
```diff
@@ -182,17 +182,11 @@ const TopicTrackingState = EmberObject.extend({
...
-      ) {
-        this.states[k] = Object.assign({}, topic, {
-          is_seen: true,
-        });
-      }
+    data.payload.topic_ids.forEach((k) => {
+      const topic = this.states[`t${k}`];
+      this.states[`t${k}`] = Object.assign({}, topic, {
+        is_seen: true,
+      });
```
Comment (lis2) [inline app/assets/javascripts/discourse/app/models/topic-tracking-state.js:189]: this logic is now simplified as message receive exact topic_ids which should be marked as "seen"

File: app/assets/javascripts/discourse/tests/unit/models/topic-tracking-state-test.js [hunk 2]
```diff
@@ -346,36 +346,7 @@ module("Unit | Model | topic-tracking-state", function (hooks) {
...
-      topic_id: 112,
-      payload: { tag_id: "foo" },
-    });
-    assert.equal(state.states["t112"].is_seen, true);
-
-    state.states["t112"].is_seen = false;
-    state.dismissNewTopic({
-      message_type: "dismiss_new",
-      topic_id: 112,
-      payload: {},
```
Comment (lis2) [inline app/assets/javascripts/discourse/tests/unit/models/topic-tracking-state-test.js:378]: This tests could be dropped as now the message is not receiving category_id or topic_id anymore

File: app/controllers/topics_controller.rb [hunk 3]
```diff
@@ -893,36 +893,30 @@ def bulk
...
-      category_ids = [params[:category_id]]
-      if params[:include_subcategories] == 'true'
-        category_ids = category_ids.concat(Category.where(parent_category_id: params[:category_id]).pluck(:id))
-      end
-
-      topic_scope =
-        if params[:tag_id]
-          Topic.where(category_id: category_ids).joins(:tags).where(tags: { name: params[:tag_id] })
-        else
-          Topic.where(category_id: category_ids)
```
Comment (lis2) [inline app/controllers/topics_controller.rb:906]: the topic scope is a filtered list to potential Dismiss. Depending on the combination of params like :category_id and :tag_id

File: app/jobs/scheduled/clean_dismissed_topic_users.rb [hunk 4]
```diff
@@ -24,7 +24,7 @@ def delete_overdue_dismissals!
                     WHEN COALESCE(user_options.new_topic_duration_minutes, :default_duration) = :always THEN users.created_at
                     WHEN COALESCE(user_options.new_topic_duration_minutes, :default_duration) = :last_visit THEN COALESCE(users.previous_visit_at,users.created_at)
                     ELSE (:now::timestamp - INTERVAL '1 MINUTE' * COALESCE(user_options.new_topic_duration_minutes, :default_duration))
-                 END, user_stats.new_since, :min_date)
+                 END, users.created_at, :min_date)
```
Comment (lis2) [inline app/jobs/scheduled/clean_dismissed_topic_users.rb:27]: I could keep here new_since as a column still existing and we will potentially get rid of that later. I wanted to already switch to use `created_at`. The reason is that after registration, the user should have an empty "new" tab. Only topics created after could be `new`

File: app/services/dismiss_topics.rb [hunk 5]
```diff
@@ -8,6 +8,7 @@ def initialize(user, topics_scope)
 
   def perform!
     DismissedTopicUser.insert_all(rows) if rows.present?
+    @rows.map { |row| row[:topic_id] }
```
Comment (lis2) [inline app/services/dismiss_topics.rb:11]: return dismissed topics ids. rows are memoized so it should not affect performance

File: app/services/dismiss_topics.rb [hunk 6]
```diff
@@ -17,6 +18,7 @@ def rows
       .joins("LEFT JOIN topic_users ON topic_users.topic_id = topics.id AND topic_users.user_id = #{@user.id}")
       .where("topics.created_at >= ?", since_date)
       .where("topic_users.id IS NULL")
+      .where("topics.archetype <> ?", Archetype.private_message)
```
Comment (lis2) [inline app/services/dismiss_topics.rb:21]: I realized that in previous commits I forgot about private messages. No big harm as DismissedTopicUsers table is regularly cleaned

File: app/services/user_merger.rb [hunk 7]
```diff
@@ -266,6 +266,8 @@ def update_user_ids
     update_user_id(:draft_sequences, conditions: "x.draft_key = y.draft_key")
     update_user_id(:drafts, conditions: "x.draft_key = y.draft_key")
 
+    update_user_id(:dismissed_topic_users, conditions: "x.topic_id = y.topic_id")
+
```
Comment (lis2) [inline app/services/user_merger.rb:270]: Another bug, merging users should move dismissed_topic_users rows

File: db/migrate/20210208022738_move_new_since_to_new_table.rb [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+      AND topics.archetype <> :private_message
+      AND topics.created_at >= GREATEST(CASE
+                  WHEN COALESCE(user_options.new_topic_duration_minutes, :default_duration) = :always THEN users.created_at
+                  WHEN COALESCE(user_options.new_topic_duration_minutes, :default_duration) = :last_visit THEN COALESCE(users.previous_visit_at,users.created_at)
+                  ELSE (:now::timestamp - INTERVAL '1 MINUTE' * COALESCE(user_options.new_topic_duration_minutes, :default_duration))
+               END, :min_date)
+      ORDER BY topics.created_at DESC
+      LIMIT :max_new_topics
+    SQL
+    sql = DB.sql_fragment(sql,
```
Comment (SamSaffron) [inline db/migrate/20210208022738_move_new_since_to_new_table.rb:26]: DB.exec will do the trick :) it accepts params

**Comment by SamSaffron** [comment]
Overall the change looks good to me. I think though we should have 1 more pair of eyes on it. @martin-brennan can you review as well.

File: db/migrate/20210208022738_move_new_since_to_new_table.rb [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+            last_visit: User::NewTopicDuration::LAST_VISIT,
+            always: User::NewTopicDuration::ALWAYS,
+            default_duration: SiteSetting.default_other_new_topic_duration_minutes,
+            min_date: Time.at(SiteSetting.min_new_topics_time).to_datetime,
+            private_message: Archetype.private_message,
+            max_new_topics: SiteSetting.max_new_topics)
+  end
+
+  def down
+    raise IrriversibleMigration
```
Comment (martin-brennan) [inline db/migrate/20210208022738_move_new_since_to_new_table.rb:37]: This is probably not necessary just thought to check, but usually this is `raise ActiveRecord::IrreversibleMigration`

**Review by martin-brennan — APPROVED** [review]
LGTM, I cannot see anything that stands out and you obviously have delved deeper into this than me :)

**Pull request merged**